app/services/foreman/plugin/rbac_support.rb

```diff
@@ -0,0 +1,76 @@
+module Foreman
+  class Plugin
+    class RbacSupport
+      def add_resource_permissions_to_default_roles(resources, opts = {})
+        # to get early warnings on typos
+        resources.map(&:constantize)
+        Role.without_auditing do
+          Role.transaction do
+            add_resource_permissions_to_viewer(resources, opts.dup)
+            add_resource_permissions_to_manager(resources, opts.dup)
+          end
+        end
+      end
+
+      def add_permissions_to_default_roles(args)
+        args.each do |key, array|
+          check_role_name_before_extending key
+          role = Role.find_by :name => key
+          next unless role
+          role.modify_locked = true
+          add_permissions_to_role(role, array)
+        end
+      end
+
+      def check_role_name_before_extending(role_name)
+        msg = "Invalid role name, only 'Viewer' of 'Manager' is allowed to be extended from a plugin"
+        raise msg if role_name != "Viewer" && role_name != "Manager"
+      end
+
+      def add_resource_permissions_to_manager(resources, opts)
+        manager = Role.find_by :name => "Manager"
+        return if !manager || resources.empty?
+        add_permissions manager, resources, opts
+      end
+
+      def add_resource_permissions_to_viewer(resources, opts)
+        viewer = Role.find_by :name => "Viewer"
+        opts[:condition] = "name LIKE :name"
+        opts[:condition_hash] = { :name => "view_%" }
+        return if !viewer || resources.empty?
+        add_permissions viewer, resources, opts
+      end
+
+      def add_permissions(role, resources, opts)
+        role.modify_locked = true
+        to_add = resources.map { |resource| new_permissions_for role, resource, opts }.reduce(:+).map(&:name)
+        role.add_permissions! to_add if to_add.any?
+      end
+
+      def new_permissions_for(role, resource, opts)
+        condition_hash = { :resource_type => resource }
+        condition_hash.merge! opts[:condition_hash] if opts[:condition_hash]
+        conditions = "resource_type = :resource_type"
+        conditions << (" AND " + opts[:condition]) if opts[:condition]
+        all_permissions = Permission.where(conditions, condition_hash)
+        all_permissions.reject do |permission|
+          permission_already_included? role, permission, opts
+        end
+      end
+
+      def permission_already_included?(role, permission, opts)
+        opts[:except] ||= []
+        opts[:except].include?(permission.name.to_sym) || role.permissions.where(:name => permission.name).any?
+      end
+
+      def add_permissions_to_role(role, permission_names)
+        Role.without_auditing do
+          to_add = permission_names.reject do |name|
```

Comment:
I think it would be better to first load all role permissions so avoid separate SQL for each permission, e.g.

```ruby
existing_permissions = role.permissions.where(:name => permission_names).pluck(:name)
to_add = permission_names.select { |name| !existing_permissions.include?(name) }
```